### ISSUE TITLE
Reverse mode run-time

### ISSUE THREAD

**Issue by davidleitao95** [autodiff/autodiff#37]
Hi,

I have an optimization problem where my cost function _f_ depends on _n_ (usually >100) input variables (vector **FA**) and also a handful of vector parameters with the same size:

`var f(const VectorXd& TR, const VectorXd& TE, VectorXvar FA, const VectorXd& RF)
{  
       //cost function calculation;
}

void cpp17_EPG_GRE_autodiff (const int& n, const VectorXd TR0, VectorXd TE0, VectorXd FA0, VectorXd RF0, double& F, VectorXd& grad)
{
        VectorXvar FA(n);
        for(int i=0; i<n; i++){
            FA(i) = FA0(i);
        }
        
     	var fval = f(TR0, TE0, FA, RF0);
        std::cout << "I'm here 1 " << val(fval) << std::endl
       
	VectorXd dydx = gradient(fval, FA);
        std::cout << "I'm here 2 " << val(fval) << std::endl;

        F = val(fval);

        for(int i=0; i<n; i++){
            grad(i) = dydx(i);
        }
}
`

For testing, I set _n=6_ variables and performed a simple function evaluation which takes <0.001 seconds, while calculating the cost function gradient takes 15 seconds. With the `std::cout` commands I can see that it is taking a lot of time at `VectorXd dydx = gradient(fval, FA);`. Is there anything wrong in my implementation from what I have shown? 

Thanks for your help,
David

**Comment by allanleal** [comment]
This is indeed a lot. I have not been able improve the reverse mode algorithm - focus has been given mainly to forward mode, which uses template meta-programming techniques to optimize the arithmetic/mathematical operations on variables of type `dual` at compile time. An the variables of `dual` type are allocated at stack memory, whereas the variables of type `var` are pointer wrappers to dynamically allocated memory. Mathematical operations on `dual` variables should then be much faster, but their purpose is different -- computing derivatives of a function with respect to individual variables (one at a time) or with respect to a direction vector (with the case of individual variables a specific case of a direction vector along the unit vector corresponding to the input variable of interest). 

The reverse mode algorithm needs to construct an expression tree at runtime (as opposed to at compile time as it happens for the forward mode). Basically, you replace operations on type `double` by type `var` which is a wrapper to a pointer to dynamic memory allocated data. The bottom line is that the reverse algorithm in `autodiff` is, currently, not optimized as the forward algorithm, and it would be great if one or more people could collaborate in this area. 

May I ask you to formulate a simple, running example, not necessary identical to your case, but that demonstrate this performance issue? I would then run it here and profile it; see where is the bottleneck.

**Comment by davidleitao95** [comment]
As my cost function links many intermediate variables each depending on the design variables, and because inheritance is used in _reverse.hpp_, I believe this is what slows down so massively my gradient calculation. I tried to build a simple example, but it seemed to work just fine in that case. It might just happen when someone links many intermediate variables that depend on the design variables. I can still send you an example where this happens but it is using my original cost function, which is very messy..

I might not use this library for my current application, but many thanks for your help!

**Comment by allanleal** [comment]
If your code is self-contained and does not depend on complicated third-party codes, then I could have a look and see how things can be improved for your specific case.

**Comment by davidleitao95** [comment]
Apologies for my delay - a lot of things have popped up to do.
Yes, my code is self-contained, only making use of eigen library.
Please feel free to ask any questions about my cost function - I'm also sorry I did not find the time to clean up my code.. 

[example-reverse-mode-run-time.txt](https://github.com/autodiff/autodiff/files/3346321/example-reverse-mode-run-time.txt)

**Comment by allanleal** [comment]
Haven't been able to run your code yet, but the cost function seems indeed complicated. Some suggestions: your objective function could the a class method, and the multiple `VectorXvar` objects it currently instantiates becoming data members of the class, to avoid their repeated allocation at every function call.

What happens is that when `f` is evaluated with `double` variables, this is blazing fast compared to `var` because an expression tree is dynamically (at runtime) created so that it can be traversed later to compute all the derivatives. I see some `toc` commands there, and I'm curious to know where is the function taking most of the time (the bottleneck).

**Comment by supersega** [comment]
@allanleal I think, that i can try to optimize code in reverse mode.

**Comment by allanleal** [comment]
That would be great @supersega !

**Comment by allanleal** [comment]
If you open a new issue, we can discuss there what needs to be done.

**Comment by supersega** [comment]
@allanleal sure!

**Comment by supersega** [comment]
#40

**Comment by davidleitao95** [comment]
> Have you tried using `dual` instead?

Yes, I did try `dual` when I first implemented it but it was similarly slow. I assumed at that time that it was because I was calculating the derivative of a scalar w.r.t. many input variables

**Comment by allanleal** [comment]
@davidleitao95 Please read this explanation I wrote in another issue (#19):
https://github.com/autodiff/autodiff/issues/19#issuecomment-506390049

Your function evaluation seems very expensive, and if you have many variables, using forward automatic differentiation will require a function call per variable to compute gradient (if a scalar function) or Jacobian (if a vector function).

Reverse mode algorithm can in principle in a single function call compute all derivatives of a scalar function (with respect to all variables). But this requires the construction of an expression tree at runtime, which can be expensive and memory consuming. 

My idea is to use forward diff to compute gradients without N function evaluations, but this is something that I can only implement and experiment in a couple of months. Forward diff uses template meta programming techniques to optimize the derivative calculations and avoid temporaries. More details in that link above.

**Comment by davidleitao95** [comment]
Thanks for alerting me to this! A couple of months might actually be when I come back to this as I have currently to finish other stuff and finite differences do the job for the time being. I can then try that new approach to calculate the derivatives.
